### DISCUSSION TITLE
Persistence - How to keep process running automatically

### DISCUSSION THREAD

**Discussion by kimjonguntastic** [abraunegg/onedrive#2746] [Q&A] [answered]
I am able to get onedrive to sync manually from my pc to the cloud. Is there a command to get it to autosync? I tried using --monitor and --synchronize and both will initiate a sync but do not provide ongoing mirroring.

Is there a modifier I need to apply to get it to keep syncing?

Also does it autorun on boot or do I need to apply a modifier to do this? 

Alternately I know I can add it to cron to get it to do what I need but first wanted to check to see if I am missing something??

**Comment by abraunegg** [comment, marked as answer]
@kimjonguntastic 
> Is there a command to get it to autosync?

Autosync is achieved by setting up a 'systemd service' which then runs the client in `--monitor` mode.

> I tried using --monitor and --synchronize and both will initiate a sync but do not provide ongoing mirroring.

I am surprised by this statement ... as `--monitor` provides *ongoing* monitoring of the local filesystem for changes, and will pull down changes done online after specific time intervals.

> Is there a modifier I need to apply to get it to keep syncing?
>
> Also does it autorun on boot or do I need to apply a modifier to do this?

This is the 'systemd' service. If you want to systemd service to run without you logging in, you need to use the `linger` capability of the platform .. but that is a Linux security aspect and has no relationship to the capabilities of this client.

> Alternately I know I can add it to cron to get it to do what I need but first wanted to check to see if I am missing something??

If you go down the 'cron' route, you want to use `--synchronize` and *not* `--monitor` and schedule this to run at a minimum every 5 minutes. Enable logging to a file so that you a logfile of actions.

Whatever you do, *do not use* `--resync` unless the application advises you to do so ....... 

**References:**
* https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#running-onedrive-as-a-system-service
* https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#how-to-run-a-user-systemd-service-at-boot-without-user-login
* https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-2/docs/usage.md#how-to-initiate-the-client-as-a-background-service
* https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-2/docs/usage.md#how-to-start-a-user-systemd-service-at-boot-without-user-login

**Reply by kimjonguntastic** [reply]
@abraunegg 
after running monitor on the command line when it finishes it doesnt return to the command line, it says sync with onedrive is complete.  To return to the command line I need to ctrl C.  Are we supposed to leave the command line open to keep it syncing?

```
onedrive --monitor --local-first --single-directory 'syncfolder'  
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Initializing monitor ...
OneDrive monitor interval (seconds): 300
Starting a sync with OneDrive
Uploading differences of syncfolder
```
```Sync with OneDrive is complete
^C^X^CGot termination signal, performing clean up
Shutting down the HTTP instance
Shutting down db connection and merging temporary data
```

**Reply by abraunegg** [reply]
> after running monitor on the command line when it finishes it doesnt return to the command line, it says sync with onedrive is complete.

Because this is *what* running using `--monitor` does .. the application does not exit. It stays active, monitoring your local file system for changes.

Please read the documentation to understand how this client operates:
* https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#using-the-client
* https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-2/docs/usage.md
* https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-2/docs/client-architecture.md

**Reply by Spiderspoon53** [reply]
--monitor - it does not exit. But it seems to when you exit terminal.
I don't know where I found the following, but it seems to have fixed the problem.
```
sudo systemctl enable onedrive@john.service
sudo systemctl start onedrive@john.service
sudo systemctl status onedrive@john.service
sudo journalctl --unit=onedrive@john -f
```
I think the first 2 lines deal with the issue and the last 2 are for information.

**Reply by abraunegg** [reply]
> --monitor - it does not exit. But it seems to when you exit terminal.

If you are running `--monitor` in a terminal window, you can use 'CTRL C' to stop the client running ... standard practice.

**Comment by Spiderspoon53** [comment]
But don't you want onedrive to keep monitoring? I don't want to keep turning it on and off again. It needs to keep running in the background. Does CTRL C stop it from running in the background?

**Reply by abraunegg** [reply]
When you are running the client *manually* in a terminal session, and using `--monitor` the client is in an endless 'monitor' loop. When you close the terminal session down, it exits the application (usually in an unclean manner).

When you issue a CTRL C event, the application understands this and attempts to shutdown the client in clean and safe manner.

> I don't want to keep turning it on and off again. It needs to keep running in the background. Does CTRL C stop it from running in the background?

To achieve the desired outcome, you can start the process as a background process. You can do this by appending a `&` to your command line.

The most appropriate way to start the client as a background process is to use the systemd service. This means that the application will start|stop based on you starting/stopping your system.
